### PR TITLE
"wake up" internal prometheus scrapper metrics (up / scrape_xxxx)

### PR DESCRIPTION
**Pull request by gillg**
**Description:**
This duplicates the target of https://github.com/open-telemetry/opentelemetry-collector/pull/2918 but in a completly different approach.
It solves https://github.com/open-telemetry/opentelemetry-collector/issues/3089 at least but also other related bugs about prometheus autogenerated metrics by its scrapper : https://prometheus.io/docs/concepts/jobs_instances/#automatically-generated-labels-and-time-series

**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector/issues/3089

Resolves :
https://github.com/open-telemetry/wg-prometheus/issues/8
Maybe https://github.com/open-telemetry/wg-prometheus/issues/41 ?

**Testing:**
No new tests because we stay standard with other metrics.
I just "wake up" dormant metrics without metadata.

**Documentation:**
Nothing more, new metrics up and scrape_xxxx will be available internaly, and so at the exporter side.

### CODE REVIEW

**Comment by gillg** [comment]
@odeke-em I was no able to contribute on your PR, and because the approach is very different I prefered create a new one to discuss with maintainers about it.

**Comment by gillg** [comment]
Metrics at exporter side after my second commit :

```
# HELP scrape_duration_seconds Duration of the scrape
# TYPE scrape_duration_seconds gauge
scrape_duration_seconds{otel_job="grafana"} 0.022050369 1620294333695
scrape_duration_seconds{otel_job="otel-collector"} 0.001391824 1620294335369
scrape_duration_seconds{otel_job="thanos-compactor"} 0.000312828 1620294327957
# HELP scrape_samples_post_metric_relabeling The number of samples remaining after metric relabeling was applied
# TYPE scrape_samples_post_metric_relabeling gauge
scrape_samples_post_metric_relabeling{otel_job="grafana"} 414 1620294333695
scrape_samples_post_metric_relabeling{otel_job="otel-collector"} 40 1620294335369
scrape_samples_post_metric_relabeling{otel_job="thanos-compactor"} 0 1620294327957
# HELP scrape_samples_scraped The number of samples the target exposed
# TYPE scrape_samples_scraped gauge
scrape_samples_scraped{otel_job="grafana"} 414 1620294333695
scrape_samples_scraped{otel_job="otel-collector"} 40 1620294335369
scrape_samples_scraped{otel_job="thanos-compactor"} 0 1620294327957
# HELP scrape_series_added The approximate number of new series in this scrape
# TYPE scrape_series_added gauge
scrape_series_added{otel_job="grafana"} 414 1620294333695
scrape_series_added{otel_job="otel-collector"} 40 1620294335369
scrape_series_added{otel_job="thanos-compactor"} 0 1620294327957
# HELP up The scraping was sucessful
# TYPE up gauge
up{otel_job="grafana"} 1 1620294333695
up{otel_job="otel-collector"} 1 1620294335369
up{otel_job="thanos-compactor"} 0 1620294327957
```

used scrape config :
```yaml
      scrape_configs:
        - job_name: 'otel-collector'
          scrape_interval: 5s
          static_configs:
            - targets: ['otel-collector:8888']
          relabel_configs:
          # Trick because otel collector not expose the job and to avoid "honor_labels" at prometheus side
          - action: replace
            replacement: otel-collector
            target_label: otel_job
        - job_name: thanos-compactor
          static_configs:
            - targets: ['172.17.0.1:10942']
          relabel_configs:
          # Trick because otel collector not expose the job and to avoid "honor_labels" at prometheus side
          - action: replace
            replacement: thanos-compactor
            target_label: otel_job
        - job_name: grafana
          static_configs:
            - targets: ['172.17.0.1:3000']
          relabel_configs:
          # Trick because otel collector not expose the job and to avoid "honor_labels" at prometheus side
          - action: replace
            replacement: grafana
            target_label: otel_job
```

**Review by dashpole — APPROVED** [review]
This is the approach I think we should take.  Just clean up some of the extra debug code you added.

**Comment by gillg** [comment]
Ok, in fact there is no "useless" code. But I would uncomment my comments to enable a logger at these points.
These logs coulbe be very useful to understand what happen. Maybe "Debug" could become "Trace".

I take any help to implement a logger

**Comment by gillg** [comment]
Fixing tests become a nightmare... I need help on receiver/prometheusreceiver/metrics_receiver_test.go ! 🆘 🙏 😅

**Comment by rakyll** [comment]
There are too many tests to fix there, @gillg. Been there, done that. Let's help you if this is the way to go.

**Review by gillg — COMMENTED** [review]
> There are too many tests to fix there, @gillg. Been there, done that. Let's help you if this is the way to go.

Thank @rakyll !
It definitely works perfectly for now, I fixed a Prometheus **exporter** test, and some other internals, but I'm a little bit confused on the current logic tests. It's complicated because new introduced metrics should be not visible in a metrics fake documents, but they count in internal metrics. So I have the feeling that we should change a little the method doCompare but I'm not sure.

I also need a little help to instance a logger and uncomment my log lines.

**Comment by rakyll** [comment]
cc @odeke-em

**Comment by dashpole** [comment]
Discussed at the prometheus wg meeting today.  One request is to keep this PR narrowly tailored to the issue at hand.  Can we only add the "up" metric for now, and defer on the other ones?  We weren't able to reach consensus as to whether the other metrics should be added during the meeting.

Until https://github.com/open-telemetry/wg-prometheus/issues/52 is resolved, we should only add the 'up' metric.

**Comment by gillg** [comment]
> Discussed at the prometheus wg meeting today. One request is to keep this PR narrowly tailored to the issue at hand. Can we only add the "up" metric for now, and defer on the other ones? We weren't able to reach consensus as to whether the other metrics should be added during the meeting.
> 
> Until [open-telemetry/wg-prometheus#52](https://github.com/open-telemetry/wg-prometheus/issues/52) is resolved, we should only add the 'up' metric.

Thank you for these informations ! I don't understand why not implement all native metrics but yes it's really simple to only fix "up". We just need to remove unwanted `case` here https://github.com/open-telemetry/opentelemetry-collector/pull/3116/files#diff-a71211e5426c3d12d9c3c0b5991e4b284568b76310438f884a37ce20655327f4R88 and it will work as previously. The metric will have no type and be rejected later in the chain.

Today the main problem is to fix test cases, and help me to instanciate a logger where it's usefull (I commited commented lines with a fake logger).
Last question, `metadata.Unit = "bool"` is ok as OTEL unit for "up" ?

**Comment by bogdandrutu** [comment]
@rakyll @Aneurysm9 @dashpole as Prometheus experts please review and help this person.

**Comment by gillg** [comment]
> @rakyll @Aneurysm9 @dashpole as Prometheus experts please review and help this person.

Thank you @bogdandrutu (and all the team 😅 ) say me if you need access to my fork, or just guide me on comments. I can do it by myself bunt I don't know the good approach to inject the logger cleanly.

About he metrics, I remove other cases than "up" and I prepare a new PR in parallel for "scrape_xxx" metrics ?

**Comment by Aneurysm9** [comment]
I also think this is the correct approach to take.  It fully resolves the compliance test issues related to the missing `up` metric and appears to be a simple solution, even if the existing tests make it somewhat painful.

I managed to successfully wrangle these tests when fixing the receiver's start time adjustment logic and expect it will require similar changes (hence the current conflict with the main branch since that PR was merged).  I've blocked off some time tomorrow to try to work on making them less awful.

As for whether it makes sense to only handle `up` in this PR, I think the work is already done to handle all of these internal metrics so we might as well carry on with that.  No point adding more work down the road when we'd undoubtedly still have to do something to deal with testing those new metrics.

**Comment by dashpole** [comment]
> As for whether it makes sense to only handle up in this PR, I think the work is already done to handle all of these internal metrics so we might as well carry on with that. No point adding more work down the road when we'd undoubtedly still have to do something to deal with testing those new metrics.

I'm also OK with that.  I figured it would be easier to have to only wrangle the tests for one metric, but if it isn't much more effort, then we can do them all and just be done.

**Comment by Aneurysm9** [comment]
I've hammered the receiver's e2e tests into a much more manageable shape.  I was going to make a PR to the source branch of this PR, but that repo doesn't seem to be accepting PRs.  For now the changes can be viewed at https://github.com/Aneurysm9/opentelemetry-collector/tree/feat/add-internal-prom-metrics.  @gillg can you pull in those changes to this branch so we can get this PR unstuck?

**Comment by gillg** [comment]
Thank you a lot @Aneurysm9 ! Good job around the tests, I lose my hairs just reading them 😅 

So last things remaining : 
* Solve the units question for up metric ("bool" ?) and scrape_xxx ("seconds" vs "s", and "count")
* Add a logger usable on metricfamily to add some debugging logs for the future
* Answer to the `//TODO convert it to OtelMetrics ?` line 68

After we are probably ok to merge

File: receiver/prometheusreceiver/internal/metricfamily.go
Comment (dashpole) [reply receiver/prometheusreceiver/internal/metricfamily.go:99]: Units appear to be specific to OpenMetrics, rather than prometheus text format.  See [prometheus/prometheus/pkg/textparse/promparse.go#L195](https://github.com/prometheus/prometheus/blob/609ba54b8f7570ba23479343eec8ba3039031322/pkg/textparse/promparse.go#L195).  For OpenMetrics, here is the documentation for units: [OpenObservability/OpenMetrics/specification/OpenMetrics.md#units-and-base-units](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#units-and-base-units).  The units we should stick to include (from the OpenMetrics link): `seconds, bytes, joules, grams, meters, ratios, volts, amperes, and celsius`.  The "up" metric probably shouldn't have units in that case.

File: receiver/prometheusreceiver/internal/metricfamily.go
Comment (dashpole) [inline receiver/prometheusreceiver/internal/metricfamily.go:107]: no unit needed here.  See the above comment.

File: receiver/prometheusreceiver/internal/metricfamily.go
Comment (dashpole) [reply receiver/prometheusreceiver/internal/metricfamily.go:103]: See the above.  We should use "seconds" here.

File: receiver/prometheusreceiver/internal/metricfamily.go
Comment (dashpole) [inline receiver/prometheusreceiver/internal/metricfamily.go:111]: no unit here.  See the above comment.

File: receiver/prometheusreceiver/internal/metricfamily.go
Comment (dashpole) [reply receiver/prometheusreceiver/internal/metricfamily.go:115]: no unit here.  See the above comment.

File: receiver/prometheusreceiver/internal/metricfamily.go
Comment (dashpole) [inline receiver/prometheusreceiver/internal/metricfamily.go:100]: The "up" metric coming from prometheus has no metadata (I queried the `api/v1/targets/metadata` endpoint on my prom server), but we have to pick something here.

It seems like the best fit might be a [StateSet](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#stateset).  WDYT @Aneurysm9?

File: receiver/prometheusreceiver/internal/metricfamily.go
Comment (dashpole) [inline receiver/prometheusreceiver/internal/metricfamily.go:70]: If I understand this comment correctly, this is tracked here: https://github.com/open-telemetry/opentelemetry-collector/issues/3137.    If that is what you are trying to capture, we can remove this TODO, and use the issue to track.

File: receiver/prometheusreceiver/internal/metricfamily.go
Comment (Aneurysm9) [reply receiver/prometheusreceiver/internal/metricfamily.go:100]: The topic of StateSets came up in the metrics data model SIG the other day and I think we still don't have a plan for supporting them in OTLP.  See https://github.com/open-telemetry/opentelemetry-specification/issues/1712.  I think for the current purpose a Gauge is fine.

File: receiver/prometheusreceiver/internal/metricfamily.go
Comment (dashpole) [inline receiver/prometheusreceiver/internal/metricfamily.go:114]: This line is probably not necessary.  It would be incredibly verbose, as it will be logged multiple times per-target per-scrape.

**Review by dashpole — APPROVED** [review]

**Review by Aneurysm9 — APPROVED** [review]

**Review by alolita — APPROVED** [review]
lgtm

**Comment by alolita** [comment]
@gillg 2 required tests are still failing (windows, linter). Can you please fix these to be able to merge. Thanks!

File: exporter/prometheusexporter/end_to_end_test.go
Comment (Aneurysm9) [inline exporter/prometheusexporter/end_to_end_test.go:160]:
```suggestion
		`test_scrape_duration_seconds 0.* .*`,
```

**Comment by gillg** [comment]
The last test fail at exporter level is not fair...

Sometimes we have `0.xxx <timestamp>`, sometimes `0 <timestamp>`, sometimes `0.xxe-xx <timestamp>`.

I thought the regex `0.* .*`  would do the job but seems not. Does a simple `.*` to accept anything after the metric name is acceptable ?

**Comment by dashpole** [comment]
maybe use `[0-9.e-]*` for the middle part, and `[0-9]+` for the final timestamp?

**Comment by gillg** [comment]
Sorry for the lack of activity, the last remaining point was the exporter test case.
I made diffrent tests but I don't understand why any regex is not working...

Now I don't know how to deal with https://github.com/open-telemetry/opentelemetry-collector/pull/3145. Does it will support natively up and scrape_ metrics ? Do I have to report my fix on it ? It introduce a lot of new tests but not with the new model of @Aneurysm9 so what is the plan ?

**Comment by alolita** [comment]
@gillg can you please fix these tests so that this PR can be merged. We're blocked on other subsequent PRs. 

@Aneurysm9 see questions above ^^

**Comment by dashpole** [comment]
I was able to rebase this, and get tests working: https://github.com/open-telemetry/opentelemetry-collector/compare/main...dashpole:internal_metrics_second.  The additional changes I had to make are in the last commit, but the rebase was the harder part...

**Review by gillg — COMMENTED** [review]
> I was able to rebase this, and get tests working: [main...dashpole:internal_metrics_second](https://github.com/open-telemetry/opentelemetry-collector/compare/main...dashpole:internal_metrics_second). The additional changes I had to make are in the last commit, but the rebase was the harder part...

Good job on rebase.... ! I updated my branch. I will take a look if something else needs to be changed.

**Comment by gillg** [comment]
Ok, without more changes it seems work as before.
What introduce all the new things about prometheus / otel conversions ? I thought this will have impacts.

**Comment by dashpole** [comment]
They are introducing the change slowly over a few PRs.  The ones they added aren't used yet.  They may need to make some changes when they rebase on this change (assuming it merges).

**Comment by Aneurysm9** [comment]
It appears that the test issues have been resolved and the current failure is a (potentially flaky) load test.  Can @open-telemetry/collector-maintainers confirm this and land this PR?

**Comment by gillg** [comment]
Data dropped due to high memory usage is common with load tests or not ?
I experienced pretty often this kind of things with my custom OTEL build. See https://github.com/open-telemetry/opentelemetry-collector/issues/3250
I think it's not related, but just in case.

**Comment by alolita** [comment]
Thanks @gillg All tests are passing finally. 🎉

@bogdandrutu can you please merge.

**Pull request merged**
